app/scripts/controllers/network/provider-api-tests/shared-tests.js

```diff
@@ -129,6 +138,377 @@ export function testsForRpcMethodAssumingNoBlockParam(method) {
+      const request = { method };
+
+      // The first time a block-cacheable request is made, the latest block
+      // number is retrieved through the block tracker first. It doesn't
+      // matter what this is — it's just used as a cache key.
+      comms.mockNextBlockTrackerRequest();
+      comms.mockInfuraRpcCall({
+        request,
+        response: {
+          httpStatus: 429,
+        },
+      });
+      const promiseForResult = withInfuraClient(async ({ makeRpcCall }) =>
+        makeRpcCall(request),
+      );
+
+      await expect(promiseForResult).rejects.toThrow(
+        'Request is being rate limited',
+      );
+    });
+  });
+
+  it('throws a custom error if the request to Infura returns a response that is not 405, 429, 503, or 504', async () => {
+    await withMockedInfuraCommunications(async (comms) => {
+      const request = { method };
+
+      // The first time a block-cacheable request is made, the latest block
+      // number is retrieved through the block tracker first. It doesn't
+      // matter what this is — it's just used as a cache key.
+      comms.mockNextBlockTrackerRequest();
+      comms.mockInfuraRpcCall({
+        request,
+        response: {
+          id: 12345,
+          jsonrpc: '2.0',
+          error: 'some error',
+          httpStatus: 420,
+        },
+      });
+      const promiseForResult = withInfuraClient(async ({ makeRpcCall }) =>
+        makeRpcCall(request),
+      );
+
+      await expect(promiseForResult).rejects.toThrow(
+        '{"id":12345,"jsonrpc":"2.0","error":"some error"}',
+      );
+    });
+  });
+
+  [503, 504].forEach((httpStatus) => {
+    it(`retries the request to Infura up to 5 times if it returns a ${httpStatus} response, returning the successful result if there is one on the 5th try`, async () => {
+      await withMockedInfuraCommunications(async (comms) => {
+        const request = { method };
+
+        // The first time a block-cacheable request is made, the latest block
+        // number is retrieved through the block tracker first. It doesn't
+        // matter what this is — it's just used as a cache key.
+        comms.mockNextBlockTrackerRequest();
+        // Here we have the request fail for the first 4 tries, then succeed
+        // on the 5th try.
+        comms.mockInfuraRpcCall({
+          request,
+          response: {
+            error: 'Some error',
+            httpStatus,
+          },
+          times: 4,
+        });
+        comms.mockInfuraRpcCall({
+          request,
+          response: {
+            result: 'the result',
+            httpStatus: 200,
+          },
+        });
+        const result = await withInfuraClient(
+          async ({ makeRpcCall, clock }) => {
+            // Note that we have to manually capture the value that the
+            // promise that `makeRpcCall` returns when it resolves or is
+            // rejected using `then`/`catch` instead of `try`/`catch` because
```

Comment:
We still don't need to use `.then` or `.catch` here. The RPC call promise can be `await-`ed at the end, after the whole loop.. I'm not sure I follow why this was done.

Reply:
This was copypasta from other similar tests where we expect the promise to be rejected rather than resolved. I've removed this in [6440071](https://github.com/MetaMask/metamask-extension/pull/15717/commits/64400716b38bd792a090ff94e6b246d904d08816).

Reply:
The same question applies for the other cases too though. Why wouldn't `await` work?

Reply:
Yeah, that's a good question. If we move the `await` at the end for the rejection cases as for the resolved cases, like so:

```
const promise = makeRpcCall(request);

while (nock.pendingMocks().length > 0) {
  // ...wait for next retry...
}

return promise;
```

then once the promise rejects — and that will happen while we are inside of the loop — we'll see a warning in the output "Promise rejection was handled asynchronously", and additionally, the error that the middleware throws will NOT be captured as a promise rejection but will kill the test right there and then. 

To make that warning go away, there needs to be a `.catch` present on that promise somewhere, but if we just put a dummy one on there and catch the real error later, like so:

```
const promise = makeRpcCall(...).catch(() => {
  // do nothing
});

// ...run loop...

return promise;
```

then the error is discarded. In fact, the whole rejection will be discarded — the test will think that this promise now resolves. So we need to capture both the resolution and rejection values asynchronously using `.then`/`.catch`, so we can monitor the pending mocks, and then once we're ready to continue we fulfill the outer promise and complete the test.

It is definitely long-winded to say it this way though :(

Reply:
Ah I see, right, I forgot about that issue. Node.js makes it difficult to deal with Promises later.

We can workaround that by adding an empty catch block. It's not pretty, but it's much more concise and lets us use `await` and pass the Promise around as we'd want to here.

e.g.

```
            const promise = makeRpcCall(request);
            promise.catch(() => {
              // Error is handled later when `promise` is awaited
            });
```

Reply:
Okay, thanks for this tip, I wasn't quite getting that 😆 

Fixed in [3924559](https://github.com/MetaMask/metamask-extension/pull/15717/commits/39245595cb95445bf9048e7509b9be76f4cb93ab).